lib/client/db/postgres/repl/repl_test.go

```diff
@@ -0,0 +1,369 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package repl
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgproto3/v2"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	clientproto "github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/client/db/postgres/repl/testdata"
+	"github.com/gravitational/teleport/lib/utils/golden"
+)
+
+func TestStart(t *testing.T) {
+	ctx := context.Background()
+	_, tc := StartWithServer(t, ctx)
+
+	// Consume the REPL banner.
+	_ = readUntilNextLead(t, tc)
+
+	writeLine(t, tc, singleRowQuery)
+	singleRowQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "single", []byte(singleRowQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "single")), singleRowQueryResult)
+
+	writeLine(t, tc, multiRowQuery)
+	multiRowQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "multi", []byte(multiRowQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "multi")), multiRowQueryResult)
+
+	writeLine(t, tc, errorQuery)
+	errorQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "err", []byte(errorQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "err")), errorQueryResult)
+
+	writeLine(t, tc, dataTypesQuery)
+	dataTypeQueryResult := readUntilNextLead(t, tc)
+	if golden.ShouldSet() {
+		golden.SetNamed(t, "data_type", []byte(dataTypeQueryResult))
+	}
+	require.Equal(t, string(golden.GetNamed(t, "data_type")), dataTypeQueryResult)
+}
+
+// TestQuery given some input lines, the REPL should execute the expected
+// query on the PostgreSQL test server.
+func TestQuery(t *testing.T) {
+	ctx := context.Background()
+	_, tc := StartWithServer(t, ctx, WithCustomQueries())
+
+	// Consume the REPL banner.
+	_ = readUntilNextLead(t, tc)
+
+	for name, tt := range map[string]struct {
+		lines         []string
+		expectedQuery string
+	}{
+		"query":                               {lines: []string{"SELECT 1;"}, expectedQuery: "SELECT 1;"},
```

Comment:
nit: ` select 1;;` or `select 1 ;  ;  `. not a big deal if we cannot handle it either.